### ISSUE TITLE
`torch.distributed.recv_object_list` doesn't move the objects to the correct device

### ISSUE THREAD

**Issue by goelayu** [pytorch/pytorch#137938]
### 🐛 Describe the bug

After the  `torch.distributed.recv_object_list(obj, dst)` method returns, the `obj` resides on the sender GPU's memory, not on the receiver GPU's memory. I would expect `obj` to be residing on the receiving GPU. 

```python
import torch
import torch.distributed as dist
import lightning as L

fabric = L.Fabric(devices=2, strategy="ddp")
fabric.launch()
    
rank = fabric.global_rank
if rank == 0:
  with fabric.init_tensor():
    tensors_send = [torch.tensor([1,2,3])
  dist.send_object_list(tensor_send, dst=1)
elif rank == 1:
  with fabric.init_tensor(): 
    tensors_recv = [torch.empty(3, dtype=torch.int))
    tensors_local = torch.tensor([7,8,9])
  dist.recv_object_list(tensor_recv,src=0)
  tensor_local = tensor_local + tensor_recv[0] //throws error since tensor_recv[0] is on cuda:0 and tensor_local on cuda:1
```

After dist.recv_object_list, why is `tensor_recv` not on the correct device (in this case, `cuda:1`). Do I need to copy tensor_recv to cuda:1 explicitly? If so, what is the point of the `.send` call?
  

### Versions

```
PyTorch version: 2.4.1+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.14 | packaged by conda-forge | (main, Mar 20 2024, 12:45:18) [GCC 12.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-121-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.5.82
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA H100 PCIe
GPU 1: NVIDIA H100 PCIe
GPU 2: NVIDIA H100 PCIe
GPU 3: NVIDIA H100 PCIe

Nvidia driver version: 555.42.06
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 57 bits virtual
Byte Order:                           Little Endian
CPU(s):                               208
On-line CPU(s) list:                  0-207
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) Platinum 8470
CPU family:                           6
Model:                                143
Thread(s) per core:                   2
Core(s) per socket:                   52
Socket(s):                            2
Stepping:                             8
CPU max MHz:                          3800.0000
CPU min MHz:                          800.0000
BogoMIPS:                             4000.00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 invpcid_single cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_pkg_req avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities
L1d cache:                            4.9 MiB (104 instances)
L1i cache:                            3.3 MiB (104 instances)
L2 cache:                             208 MiB (104 instances)
L3 cache:                             210 MiB (2 instances)
NUMA node(s):                         2
NUMA node0 CPU(s):                    0-51,104-155
NUMA node1 CPU(s):                    52-103,156-207
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] bert-pytorch==0.0.1a4
[pip3] clip-anytorch==2.6.0
[pip3] CoCa-pytorch==0.1.0
[pip3] dalle2-pytorch==1.14.2
[pip3] ema-pytorch==0.4.8
[pip3] functorch==1.14.0a0+b71aa0b
[pip3] mypy==1.10.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.23.5
[pip3] onnx==1.16.1
[pip3] open-clip-torch==2.24.0
[pip3] pytorch-labs-segment-anything-fast==0.2
[pip3] pytorch-lightning==2.4.0
[pip3] pytorch-warmup==0.1.1
[pip3] rotary-embedding-torch==0.3.3
[pip3] torch==2.4.1
[pip3] torch-fidelity==0.3.0
[pip3] torch_geometric==2.4.0
[pip3] torchao==0.5.0
[pip3] torchaudio==2.4.1
[pip3] torchmetrics==1.4.0.post0
[pip3] torchmultimodal==0.1.0b0
[pip3] torchpippy==0.2.0
[pip3] torchvision==0.19.1
[pip3] triton==3.0.0
[pip3] vector-quantize-pytorch==1.14.24
[conda] pytorch-lightning         2.4.0                    pypi_0    pypi
[conda] torch                     2.4.1                    pypi_0    pypi
[conda] torchao                   0.5.0                    pypi_0    pypi
[conda] torchaudio                2.4.1                    pypi_0    pypi
[conda] torchpippy                0.2.0                    pypi_0    pypi
[conda] torchvision               0.19.1                   pypi_0    pypi
[conda] triton                    3.0.0                    pypi_0    pypi
```

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

**Comment by awgu** [comment]
Similar to https://github.com/pytorch/pytorch/issues/61591

**Comment by fduwjj** [comment]
@awgu Are you saying this is also caused by `_unpickler(io.BytesIO(buf)).load()` inside `_tensor_to_object` which load to the wrong device? Would set device actually helps this issue?

**Comment by wconstab** [comment]
I think this is actually not a bug.  You may be using the wrong API for what you want to do.

dist.send/recv APIs would handle the case of sending tensor data.  And if you send a tensor that was on gpu0 from rank0 to rank1, rank1 would recv the tensor on the device specified by the recv call.

send_object/recv_object are intended to help send python objects by smuggling them as tensor data.  Roughly the objects are pickled, the size of the pickle is determined, a byte-tensor is created with the pickle data, and then send/recv operations are performed to send the raw bytes.  On the recv side, unpickle is used.  If one of these objects is a torch.Tensor, it gets pickled faithfully - including which device it's on.  On the reconstruction side, unpickling will find the device metadata and use that to construct a faithful copy of the original.

But it's not recommended to use object APIs for sending/receiving tensor data anyways.  Tensor data is slow to pickle and unpickle, and this step is skipped when using the regular send/recv APIs.

**Comment by goelayu** [comment]
@wconstab regardless of how the API is used, I believe `dist.recv_object_list` should have the same semantics as `dist.recv`, in the sense that the received tensor/objects are on the device specified by the method. 

Also in the code snippet I provided, I used tensors in place of Python objects simply for demonstration purposes. In reality I am using the `send_object_list` API to send objects.

**Comment by wconstab** [comment]
Semantically the apis are different.

Recv is tensor specific and knows about device and lets you control the recv tensor.

The object apis are decidedly un-specialized and know nothing about tensors or devices of the underlying objects. You could include any objects in there, and the object api won't treat any of them specially other than calling pickle/unpickle. So again, not recommended to use the object apis for tensors. 

So the question is what are the semantics of pickling a tensor on one device, then in another process unpicking it? The answer, IIUC, is the unpickler recreates the original tensor including its device.

**Comment by goelayu** [comment]
Thanks for the explanation. Maybe updating the docs would be helpful, to reflect this. 

Looking at the [docs](https://pytorch.org/docs/stable/distributed.html#torch.distributed.recv_object_list), `recv_object_list` takes a `device` argument, which is described as: "If not None, receives on this device.". I believe most would interpret this to be that the object would reside on the `device`'s memory

**Comment by goelayu** [comment]
Thinking more about this, I was wondering if there was a way to serialize/deserialize the data on the GPU itself. That could reduce the data-path significantly as no more sending data to the CPU for any GPU-GPU p2p transfers.

**Comment by fduwjj** [comment]
> That could reduce the data-path significantly as no more sending data to the CPU for any GPU-GPU p2p transfers.

@goelayu regarding this, is there any reason you cannot use send/recv directly which is designed for tensor? I mean do you want pass around non-tensor from one GPU to another GPU? IIUC, usually people leverage GPU for tensor related computations and speedup, I am curious what non-tensor you want to pass between GPUs, could you please elaborate on that?

**Comment by goelayu** [comment]
@fduwjj  I am trying to send model shards from one GPU to another. I guess one way to do that would be to iterate through all the parameters and individually send them to the other GPU and then reconstruct them on there? Would be a lot more work, but also performance would be really bad?

**Comment by wconstab** [comment]
Iteratively sending parameter tensors might give optimal performance if the parameters are not too small. If they are small enough that the overhead of individual sends is a problem, you can either try batch-send or you can torch.cat the individual tensors into one larger tensor and send that. Any of these should perform better than using object APIs, which add a huge overhead of pickling the tensor data.

I agree we should update the docs to make it more clear what the object APIs are for. To your point, the specified device is actually used for receiving the data, but all the objects end up being passed back to cpu for unpickling and reconstruction. The api is really designed for python objects not tensors. We should clarify this in our doc.

**Comment by goelayu** [comment]
Thanks. Turns out the biggest issue wasn't even the poor performance of the object API, it was the incredibly high latency of the very first NCCL API call. For some reason NCCL warmup is 3-4 orders of magnitude slower than the subsequent calls. 

If anyone is interested in the precise numbers, I am sending about 500MB from one H100 to H100 on the same node, connected via NVLink.
```
Including NCCL warmup (with object APIs): 1.6s
Excluding NCCL warmup (with object APIs): 2ms
Excluding NCCL warmup (with tensor APIs): 10us
```
